Various TS infantry fixes

reaperrr opened the pull request:
- fixed zapped animation colors
- cut down a lot of redundancy in sequences (I recommend looking at the full file, not just the diff)
- implemented TS-style Cyborg death explosions
- fixed crippled cyborgs returning to full speed once regaining health

The last remaining bit to make cyborgs work like in TS would be to play an explosion on the normal -> crippled transition, but I need more time to come up with something non-hacky.

reaperrr committed:
Refactored TS infantry sequences

They now make heavy use of inheritance to cut down duplication. Most infantry in TS uses the Minigunner (E1) sequences, only that some frames are empty (civ2 shoot animation, for example) or used for a different purpose (panic anims for civilians, crippled legs for cyborgs).

To make this inheritance a bit easier, the attack sequences were renamed to a generic 'attack', rather than 'shoot' or 'throw', and split off to a separate abstract default inherited only by infantry with valid attack frames.

reaperrr committed:
Remove obsolete custom electro.shp

This shp was originally used to work around a palette issue with the original file inside conquer.mix, but since then we actually made it use the correct palette, without removing this converted shp. In other words, removing this file fixes the colors of the Zapped infantry death.

reaperrr committed:
Make Cyborgs explode regardless of death type

Also split out death-related traits of all non-cyborg infantry to a separate inheritable default.

reaperrr committed:
Allow UpgradeOnDamage to grant permanent upgrades

Needed to properly implement things like permanently crippled weapons or limbs.

reaperrr committed:
Fix Cyborgs to stay slow after losing their legs

Even if regaining health.

pchote commented:
> The last remaining bit to make cyborgs work like in TS would be to play an explosion on the normal -> crippled transition, but I need more time to come up with something non-hacky.

Just write a trait that implements `INotifyDamageStateChanged` and fires a weapon.  The upgrades stuff didn't *completely* supersede the ability to write special-case traits :wink:

GraionDilach commented on mods/ts/sequences/infantry.yaml:
```diff
@@ -45,110 +47,263 @@ e1.gdi:
+		Facings: 8
+		Stride: 9
+		ShadowStart: 418
 	die1:
```
What's the point in keeping/editing these? You removed WithDeathAnimation entirely from cyborgs with the new template.

reaperrr replied:
I just forgot to remove them, tbh.

GraionDilach commented:
Regarding infdie.shp and it's removal from HeadshotDeath: I'm fine with that, but IMO purge HeadshotDeath entirely then, IIRC it's only used at Umagon - remember http://logs.openra.net/?year=2015&month=07&day=18#22:17:12 - This could also be an opportunity to restore die3 into ExplosionDeath properly (s_bang34) although I dunno if a weapon in TS actually used that deathtype.

reaperrr committed:
Introduce ExplosionOnDamageTransition

And use it for TS cyborgs.

pchote commented on mods/ts/sequences/infantry.yaml:
```diff
@@ -1,5 +1,5 @@
-e1.gdi:
-	Defaults: e1
+^e1:
```
Can we call this `^BasicInfantry` or something similar?

pchote commented on mods/ts/sequences/infantry.yaml:
```diff
@@ -45,110 +47,263 @@ e1.gdi:
+	die6: electro
+		Length: *
+
+^e1attack:
```
Likewise here.

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -339,8 +304,50 @@
+		Probability: 10
+		OwnerType: InternalName
+		InternalOwner: Creeps
+		DeathType: EnergyDeath # TODO: FIX ME! (Tiberium currently uses the wrong damage type!)
```
Wouldn't this be a great time to fix that?

reaperrr committed:
Remove TS Headshot DeathType

Also fixed a couple of mistakes compared to the original (some weapons still applied the wrong death type).

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -339,8 +304,50 @@
+^RegularInfantryDeath:
+	WithDeathAnimation@normal:
+		DeathTypes:
+			BulletDeath: 1
```
Could we change these to use descriptive death suffixes instead of integers? I'm assuming that you only need to change a handful of places now that you've deduplicated both the rules and sequence setup.

reaperrr committed:
Fix Tiberium and Vein DamageTypes

reaperrr committed:
Make TS inf death sequences descriptive

reaperrr committed:
Change WithDeathAnimation to look up string rather than integers

For DeathType sequence mapping.

reaperrr commented:
Updated.

reaperrr commented:
Updated again.

GraionDilach commented on mods/ts/rules/defaults.yaml:
```diff
@@ -339,8 +304,52 @@
+		CrushedSequencePalette: player
+		CrushedPaletteIsPlayerPalette: true
+	DeathSounds@NORMAL:
+		DeathTypes: BulletDeath, SmallExplosionDeath, ExplosionDeath
```
You sure about this? Shouldn't ExplosionDeath have one of the expnew auds as it's sound?

reaperrr replied:
It should have both, actually. I just played the original and there was a death scream in addition to the explosion. I'll leave this entry as-is and add a `DeathSounds@EXPLOSION` that only triggers on `ExplosionDeath`.

reaperrr committed:
Add explosion sound to TS explosion inf death

reaperrr commented:
Updated.

GraionDilach commented:
:+1:

abcdefg30 commented on OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamage.cs:
```diff
@@ -29,26 +29,32 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		[Desc("Levels of damage at which to grant upgrades.")]
 		public readonly DamageState ValidDamageStates = DamageState.Heavy | DamageState.Critical;
 
+		[Desc("Are upgrades irrevokable once the conditions have been met?")]
```
`irrevokable` -> `irrevocable`

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

Mailaender commented:
debug.log now spams:

```
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantry.shp
File not found: ^BasicInfantryAttack.shp
File not found: ^BasicInfantryAttack.shp
```

reaperrr commented:
I might be mistaken, but to me it looks like this is rather our engine's fault for not recognizing those as abstract sequences, rather than this PR's.

Mailaender commented:
That wasn't meant as a reproach. It just needs to be fixed one way or the other.
